### ISSUE TITLE
[BUG?] Strided transposed convolution returning wrong shape

### ISSUE THREAD

**Issue by eickenberg** [pytorch/pytorch#5057]
Hi everybody,

`conv_transpose2d` and `conv_transpose1d` seem to be broken for certain configurations of shapes and strides - IFF by 'conv_transpose' we agree that we mean to compute the transpose/adjoint of the convolution operation with the same parameter specs. In this case, the transpose has to verify `dot(Ax, y) = dot(x, A.Ty)` for any `x, y`, where `A` is the convolution operator and the `dot`s are inner products.

Here is a test that can be run for different configurations (code is copy+pasteable):

```python
import numpy as np
import torch
from torch.autograd import Variable
from torch.nn import functional as F
from sklearn.utils import check_random_state



def test_convolution_adjointness(signal_spatial_shape, kernel_spatial_shape, stride, 
                                 conv=F.conv2d, conv_transpose=F.conv_transpose2d, random_state=42,
                                 n_samples=5, n_in_channels=2, n_out_channels=3):
    """This test checks whether the operations conv_*d and conv_transpose_*d are actually tranposed/adjoint
        to each other. This is important for exact gradient computations and to know exactly what operations
        we are dealing with. The test checks the definition for transpose in spaces with inner product:
        
        If A is the linear map in question, and `dot_d` is the inner product in its domain and `dot_c` is the
        inner product in its codomain (call both of them `dot`, because 1. it is clear which one is meant and
        2. we use the canonical one always anyway), then its transpose A.T satisifies
        
        dot(Ax, y) = dot(x, A.Ty)
        
        for any vectors x and y chosen from the domain and codomain of A.
    """

    # using numpy rng for local random seed fixing
    rng = check_random_state(random_state) 

    # Prepare random signal and kernel
    signal_shape = (n_samples, n_in_channels) + signal_spatial_shape
    kernel_shape = (n_out_channels, n_in_channels) + kernel_spatial_shape
    
    signal = Variable(torch.from_numpy(rng.randn(*signal_shape).astype('float32')))
    kernel = Variable(torch.from_numpy(rng.randn(*kernel_shape).astype('float32')))
    
    # convolve to get convolution output and its shape
    convolved = conv(signal, kernel, stride=stride)
    
    # make the random vector y for the dot products
    random_codomain_vector = Variable(torch.from_numpy(
            rng.randn(*convolved.size()).astype('float32')))
    
    # compute the codomain dot product dot(Ax, y)
    codomain_dot = (convolved * random_codomain_vector).resize(
        n_samples, int(np.prod(convolved.size()[1:]))).sum(dim=1)
    
    # transposed convolution
    transposed_output = conv_transpose(random_codomain_vector, kernel, stride=stride)
    
    # compute domain dot product dot(x, A.Ty)
    domain_dot = (transposed_output * signal).resize(
        n_samples, int(np.prod(signal.size()[1:]))).sum(dim=1)
    
    np.testing.assert_array_almost_equal(domain_dot.data.cpu().numpy(), codomain_dot.data.cpu().numpy(), decimal=3)
```

The test passes for
```python
test_convolution_adjointness((32, 32), (3, 3), 1)
test_convolution_adjointness((31, 31), (4, 4), 1)

test_convolution_adjointness((31, 32), (5, 4), 1)
test_convolution_adjointness((30, 31), (5, 4), 1)

test_convolution_adjointness((32, 32), (4, 4), 1)
test_convolution_adjointness((32, 32), (4, 4), 2)
test_convolution_adjointness((32, 32), (4, 4), 4)

test_convolution_adjointness((31, 31), (3, 3), 1)
test_convolution_adjointness((31, 31), (3, 3), 2)
test_convolution_adjointness((31, 31), (3, 3), 4)

test_convolution_adjointness((31, 32), (3, 4), 1)
test_convolution_adjointness((31, 32), (3, 4), 2)
test_convolution_adjointness((31, 32), (3, 4), 4)

```

It looks like it always passes when the stride is 1. Further, it passes when signal axis and filter axis are both of same parity (either both odd or both even), and the stride for 32 signal size is a power of 2 less than 8.

And it fails for
```python
test_convolution_adjointness((32, 32), (3, 3), 2)
test_convolution_adjointness((31, 31), (4, 4), 2)
test_convolution_adjointness((32, 32), (4, 3), 2)
test_convolution_adjointness((31, 31), (4, 3), 2)
test_convolution_adjointness((32, 32), (3, 3), 4)

test_convolution_adjointness((32, 32), (4, 4), 3)

test_convolution_adjointness((32, 32), (4, 4), 8)
```

This is when signal and filter size are not of same parity, or the stride is not of the same parity or the stride is too big.

All the failures are shape-related: The output shape is off by at least 1.

Similar things hold for `conv_transpose1d`. I did not check for 3d. Can anybody reproduce this?

[[Question: Are convolution gradients computed by the same underlying mechanism? (I'm new to pytorch and don't understand the low-level code enough yet to find out. If so, this would have some implications, in case this error persisted there and passed through the meshes due to torch shape management. The recent use of 4x4 filters might make this not be an issue, though)]]

specs:
ubuntu 16.04
pytorch 0.3.0 (conda version. persists on other versions)
1080Ti (but using cpu here)

**Comment by SsnL** [comment]
In fact the output and input shapes of convs are many-to-one, given the input parameters. For example, in `test_convolution_adjointness((32, 32), (3, 3), 2)`, both `(32, 32)` and `(31, 31)` map to `(15, 15)`. That's why we have the `output_padding` option for conv transpose. I tried adding `output_padding=1` for this case and the test passes. :)

**Comment by eickenberg** [comment]
Ah yes, good point. Makes sense.
If the padding is applied before the convolution (which is what I think the
docs are saying)

Thanks!
